modules/auxiliary/gather/ldap_hashdump.rb

```diff
@@ -0,0 +1,395 @@
+        # where the left part is a salt and the right part is MD5(Salt+NTHash)
+        # attribute value may contain multiple concatenated history entries
+        # for john sort of 'md5($s.md4(unicode($p)))' - not tested
+        hash_format = 'sambapasswordhistory'
+      when 'krbprincipalkey'
+        hash_format = 'krbprincipal'
+        # TODO: krbprincipalkey is asn.1 encoded string. In case of vmware vcenter 6.7
+        # it contains user password encrypted with (23) rc4-hmac and (18) aes256-cts-hmac-sha1-96:
+        # https://github.com/vmware/lightwave/blob/d50d41edd1d9cb59e7b7cc1ad284b9e46bfa703d/vmdir/server/common/krbsrvutil.c#L480-L558
+        # Salted with principal name:
+        # https://github.com/vmware/lightwave/blob/c4ad5a67eedfefe683357bc53e08836170528383/vmdir/thirdparty/heimdal/krb5-crypto/salt.c#L133-L175
+        # In the meantime, dump the base64 encoded value.
+        hash = Base64.strict_encode64(hash)
+      when 'userpkcs12'
+        # if we get non printable chars, encode into base64
+        if (hash =~ /[^[:print:]]/).nil?
+          hash_format = 'pkcs12'
+        else
+          hash_format = 'pkcs12-base64'
+          hash = Base64.strict_encode64(hash)
+        end
+      else
+        if hash.start_with?(/{crypt}.?\$1\$/i)
+          hash.gsub!(/{crypt}.{,2}\$1\$/i, '$1$')
+          hash_format = 'md5crypt'
+        elsif hash.start_with?(/{crypt}/i) && hash.length == 20
+          # handle {crypt}traditional_crypt case, i.e. explicitly set the hash format
+          hash.slice!(/{crypt}/i)
+          hash_format = 'descrypt' # FIXME: what is the right jtr_hash - des,crypt or descrypt ?
+        # identify_hash returns des,crypt, while JtR acceppts descrypt
+        else
+          # handle vcenter vmdir binary hash format
+          if hash[0].ord == 1 && hash.length == 81
+            _type, hash, salt = hash.unpack('CH128H32')
+            hash = "$dynamic_82$#{hash}$HEX$#{salt}"
+          else
+            # Remove LDAP's {crypt} prefix from known hash types
+            hash.gsub!(/{crypt}.{,2}(\$[0256][aby]?\$)/i, '\1')
+          end
+          hash_format = identify_hash(hash)
+        end
+      end
+
+      # higlight unresolved hashes
+      hash_format = '{crypt}' if hash =~ /{crypt}/i
+
+      print_good("#{peer} Credentials (#{hash_format.empty? ? 'password' : hash_format}) found in #{attr}: #{dn}:#{hash}")
+
+      # known hash types should have been identified,
+      # let's assume the rest are clear text passwords
+      if hash_format.nil? || hash_format.empty?
+        create_credential(service_details.merge(
+          username: dn,
+          private_data: hash,
+          private_type: :password
+        ))
+      else
+        create_credential(service_details.merge(
+          username: dn,
+          private_data: hash,
+          private_type: :nonreplayable_hash,
+          jtr_format: hash_format
+        ))
+      end
```

Comment:
I think the reporting here is incomplete. I think what you want is:
```suggestion
      if hash_format.nil? || hash_format.empty?
        credential = create_credential(service_details.merge(
          username: dn,
          private_data: hash,
          private_type: :password
        ))
      else
        credential = create_credential(service_details.merge(
          username: dn,
          private_data: hash,
          private_type: :nonreplayable_hash,
          jtr_format: hash_format
        ))
      end

      create_credential_login({
        core: credential,
        access_level: 'User',
        status: Metasploit::Model::Login::Status::UNTRIED
      }.merge(service_details))
```

Without the `create_credential_login` call, the host and service fields are not populated in the `creds` output, and the `creds -d` command (to delete all credentials for another round of testing) fails with a stack trace.

Reply:
Thanks, actually I wasn't sure, what's the purpose of 'create_credentail_login'.